### PR TITLE
Multiple devices

### PR DESCRIPTION
**Pull request by SergeyTsimfer**
Added support for multiple devices. To use this feature, one must include `device` key in the config:
```
config = {'device': ['gpu:0', 'gpu:2']}
```
Regular expressions are also allowed (e.g. 'gpu:*'). If no config option is supplied, then only one device is used.
Under the hood, there is a copy of model on every device, and each batch of data is split (along batch dimension) evenly between them. Every part is sent to separate device, where gradients are computed, then  combined on the leading one (the first device in the config), and, finally, applied to models on each device.

### CODE REVIEW

**Comment by SergeyTsimfer** [comment]
TODO:
Use `store_to_attr`/`get_from_attr` in `get_tensor_config` for convinience. Also, we can use `_check_tensor` there. 
Manually check that helper-functions like `num_channels`, `data_format` etc work as expected.
Fix collateral damage to inherited classes.

File: batchflow/models/tf/base.py [hunk 1]
```diff
@@ -242,76 +245,153 @@ def __init__(self, *args, **kwargs):
...
-            return context
+                self._attrs[attr][device] = graph_item
+
+    def get_from_attr(self, attr, device=None, default=None):
+        """ Get item from private container or directly from model graph."""
+        device = device or self.leading_device
+        if attr in self._attrs:
+            if isinstance(self._attrs[attr], dict):
+                if device in self._attrs[attr]:
+                    graph_item = self._attrs[attr][device]
```
Comment (roman-kh) [inline batchflow/models/tf/base.py]:
```suggestion
                    return self._attrs[attr][device]
```
It's `python` so you can write a shorter code

File: batchflow/models/tf/base.py [hunk 2]
```diff
@@ -242,76 +245,153 @@ def __init__(self, *args, **kwargs):
...
                 with tf.variable_scope('globals'):
-                    if self.is_training is None:
-                        self.store_to_attr('is_training', tf.placeholder(tf.bool, name='is_training'))
-                    if self.global_step is None:
-                        self.store_to_attr('global_step', tf.Variable(0, trainable=False, name='global_step'))
-
-                config = self.build_config()
-                self._full_config = config
-                self._build(config)
+                    is_training = tf.placeholder(tf.bool, name='is_training')
```
Comment (roman-kh) [inline batchflow/models/tf/base.py]: You don't seem to use it anywhere else?

File: batchflow/models/tf/base.py [hunk 3]
```diff
@@ -547,73 +633,94 @@ def _make_mip(self, input_name, tensor, config, depth):
 
     def to_classes(self, tensor, input_name, name=None):
         """ Convert tensor with labels to classes of ``input_name`` """
-        if tensor.dtype in [tf.float16, tf.float32, tf.float64]:
+        if tensor.dtype in [tf.int16, tf.int32, tf.int64, tf.float16, tf.float32, tf.float64]:
```
Comment (roman-kh) [inline batchflow/models/tf/base.py]: Why is this check needed?

File: batchflow/models/tf/base.py [hunk 4]
```diff
@@ -711,27 +822,56 @@ def _make_scope(self, config):
...
+            expanded = [tf.expand_dims(g, 0) for g, _ in grad_and_vars if g is not None]
+            concatted = tf.concat(expanded, axis=0)
+            averaged = tf.reduce_mean(concatted, axis=0)
+
+            # Apply gradient on the leading device, then distribute to the others
+            leading_device_variable = grad_and_vars[0][1]
+            apply_op = optimizer.apply_gradients([(averaged, leading_device_variable)],
+                                                 global_step=self.get_from_attr('global_step'))
+
+            distribute_weights = [v.assign(leading_device_variable)for _, v in grad_and_vars[1:]]
```
Comment (roman-kh) [inline batchflow/models/tf/base.py]:
```suggestion
            distribute_weights = [v.assign(leading_device_variable) for _, v in grad_and_vars[1:]]
```

File: batchflow/models/tf/base.py [hunk 5]
```diff
@@ -1054,12 +1208,40 @@ def save(self, path, *args, **kwargs):
...
+        if isinstance(graph_item, tf.Operation):
+            return ('Operation', graph_item.name)
+        if isinstance(graph_item, tf.Variable):
+            return ('Variable', graph_item.op.name)
+        if isinstance(graph_item, (bool, str, int, float)) or graph_item is None:
+            return graph_item
+
+        # Handle different containers
+        if isinstance(graph_item, (list, tuple)):
+            return [self._to_names(item) for item in graph_item]
```
Comment (roman-kh) [inline batchflow/models/tf/base.py]:
```suggestion
            return type(graph_item)([self._to_names(item) for item in graph_item])
```

File: batchflow/models/tf/base.py [hunk 6]
```diff
@@ -287,16 +284,11 @@ def get_from_attr(self, attr, device=None, default=None):
...
-            else:
-                graph_item = self._attrs[attr]
-        elif default is not None:
-            graph_item = default
-        else:
-            graph_item = self._check_tensor(attr, device)
-        return graph_item
+                    return self._attrs[attr][device]
+            return self._attrs[attr]
+        if default is not None:
```
Comment (roman-kh) [inline batchflow/models/tf/base.py]: This make impossible for `default` to be `None`.

File: batchflow/models/tf/encoder_decoder.py [hunk 7]
```diff
@@ -124,7 +124,7 @@ def default_config(cls):
 
     def build_config(self, names=None):
         config = super().build_config(names)
-        config['head']['targets'] = self.targets
+        config['head']['targets'] = self.get_from_attr('targets')
```
Comment (roman-kh) [inline batchflow/models/tf/encoder_decoder.py]:
```suggestion
        config['head/targets'] = self.get_from_attr('targets')
```

File: batchflow/models/tf/gcn.py [hunk 8]
```diff
@@ -57,7 +57,7 @@ def build_config(self, names=None):
         config = super().build_config(names)
         if config.get('head/num_classes') is None:
             config['head/num_classes'] = self.num_classes('targets')
-        config['head']['targets'] = self.targets
+        config['head']['targets'] = self.get_from_attr('targets')
```
Comment (roman-kh) [inline batchflow/models/tf/gcn.py]:
```suggestion
        config['head/targets'] = self.get_from_attr('targets')
```

File: batchflow/models/tf/linknet.py [hunk 9]
```diff
@@ -57,7 +57,7 @@ def build_config(self, names=None):
             config['body/filters'] = 2 ** np.arange(config['body/num_blocks']) * config['initial_block/filters']
         if config.get('head/num_classes') is None:
             config['head/num_classes'] = self.num_classes('targets')
-        config['head']['targets'] = self.targets
+        config['head']['targets'] = self.get_from_attr('targets')
```
Comment (roman-kh) [inline batchflow/models/tf/linknet.py]:
```suggestion
        config['head/targets'] = self.get_from_attr('targets')
```

**Comment by roman-kh** [comment]
@Pennyroyall, fix conflicts, please
